packages/lesswrong/server/resolvers/revisionResolvers.ts

```diff
@@ -160,6 +160,25 @@ defineQuery({
   },
 });
 
+defineQuery({
+  name: "latestGoogleDocMetadata",
```

Comment:
Seems somewhat too complicated / not very idiomatic to me. I'd add a resolver only field that got the latest revision google doc id, and didn't bother with the version.

Reply:
I agree it's a bit unidiomatic, I did think about adding a resolver only field but ended up deciding on this way for a couple of reasons. Separating out the "why make it a separate query" and "why support the version" aspects.

**Why make it a separate query?**

This component doesn't have access to the post object already, so making this a resolver only field would require either:
1. Adding a useSingle for the post, and making sure to use the same fragment as PostsEditForm (which is PostsEditQueryFragment) so it hits the cache
2. Adding a useSingle and creating a new fragment which just contains this field (plus probably PostsMinimumInfo)

I would prefer (1), but this is fragile as it introduces undeclared coupling between PostsEditForm and this file, which is likely to break if either change. Also, I wanted this to return before the whole post body loads (which can take a while), which wouldn't be possible with (1) because it relies on waiting for the main post query to return.

(2) would be fine but I still think it's more costly than this solution in terms of complexity/performance. We would still have to add this resolver function and a new fragment, and it would have to fetch the whole post on the server before getting to the resolver only field.

BEGIN graphql_rant

Btw I think this is a good example of how "graphql brain" can lead to bloat in data returned + overall complexity. It makes it very easy to throw new fields in big existing queries and reuse these for different purposes, and it makes it inconvenient to fetch a single piece of data. In the case of (2) we would have to fetch the whole post object (including the slow contents field) in order to return this one string that doesn't actually depend on the post itself. In (1) we would be adding some caching voodoo in order to work around this performance hit (hence how this comes out in terms of complexity).

Moving contents off posts at another good example of this complexity cost imo. We put contents _on_ posts because by default graphql adds a big cost to doing the JOIN onto revisions to include just that field, and now we're taking it _off_ posts now we've fixed the join thing because graphql makes it hard to _not_ include the field if we don't want it (the query to the db happens via magic that makes it hard to select out certain fields). (to some extent this is mongo/vulcan's fault too)

If we were using e.g. [tRPC](https://trpc.io/) or just a regular REST api from the beginning it would be more idiomatic by default to just fetch this one piece of data, or to add it to another api call that returns the bulk of the data on the page (as in (1)), but where this api call would likely be faster to begin with (bc less bloated) so this would be less of a problem.

END graphql_rant

**Why support the version?**

I'm somewhat predicting "versions will be a bit more important now" as a result of the other updates related to google doc importing*. Also I think given this is done as a query it doesn't add much complexity, if it were a resolver only field then maybe it would make less sense.

* Ability to load versions without publishing them, plus the nudge in the google doc import box to use version history. Plus the motivation for adding these in the first place which is that I thought google doc importing would make different versions a lot more salient (because there are now these big changes where you re-fetch your google doc)